mantidproject/mantid
# Abins: import force constants data using Euphonic

**Description of work.**

Builds on (and no longer blocked by) #30760 

Use the new semi-optional Euphonic dependency to allow import from force-constant data in .castep_bin (CASTEP), .yaml (Phonopy) or .json (Euphonic native) formats.

This greatly expands the range of atomistic codes that could be used with Abins (via Phonopy) and lays the groundwork for implementing more sophisticated simulation schemes that properly account for energy-q relationships.

**To test:**
*Manual testing*

See #30760 testing guide for some suitable input files.

Open Abins Algorithm, select ForceConstants as input file type and run with .castep_bin, .yaml etc. files.
(For Euphonic .json try something from the Euphonic *tests_and_analysis/test/data/force_constants* directory.)
If Euphonic is not installed, the VibrationalOrPhononFile field should fail to validate, with a tooltip advising to install Euphonic with an installer from the ScriptRepository. (This should not be necessary for Conda-built Mantid.)

The q-point sampling mesh is determined by a variable in `abins.parameters`. This setup is used somewhat consistently in Abins to allow "advanced" options to be tweaked. The default value is on the high side to be safe, but  the best-practice would be to check convergence. (This is not an option for other formats which use pre-computed phonon modes.) If Abins is re-run after changing `abins.parameters.sampling['force_constants']['qpt_cutoff']`, the Abins log should show that this invalidated the cache, and a new set of scattering spectra is calculated. The Monkhorst-Pack grid (i.e. regular divisions of the reciprocal unit cell) is related to the cutoff; the sampling is reported in the log and each direction should scale linearly with the cutoff and inversely with the crystalline unit cell size. (Values are rounded up so this behaviour may be less clear for large cells.)

*Unit testing*
At the moment there is no automatic unit test for this feature. This is because Euphonic is not yet available on the unit-testing CI. I propose to revisit this when the CI is running on a Conda build. But perhaps it would be an idea to include some kind of skippable test before then, which runs if Euphonic is available? It would be easy enough to hack something, but do we have a standard/maintainable way of doing that?

There is no associated issue.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
